client/signup/steps/user/index.jsx

```diff
@@ -450,12 +451,21 @@ export class UserStep extends Component {
 			query.redirect_to = window.sessionStorage.getItem( 'signup_redirect_to' );
 			window.sessionStorage.removeItem( 'signup_redirect_to' );
 		}
-		this.submit( {
-			service,
-			access_token,
-			id_token,
-			userData,
-			queryArgs: query,
+
+		const socialInfo = {
+			service: service,
+			access_token: access_token,
+			id_token: id_token,
+		};
+
+		this.props.loginSocialUser( socialInfo, '' ).finally( () => {
+			this.submit( {
+				service,
+				access_token,
+				id_token,
+				userData,
+				queryArgs: query,
+			} );
```

Comment:
Hi @heavyweight :wave: Since this PR was merged 3 months ago, we get very frequent errors in Sentry where `loginSocialUser` returns a rejected promise and that promise is not handled. `.finally` is not considered as handling apparently.

The error is always like
```
{
  code: 'unknown_user',
  message: 'Hmm, we can't find a WordPress.com account for that social login.'
}
```
and the path is `/start/user-social`. Do you have any idea how to fix this? If the social login failed, we don't really want to submit the `user-social` step, do we? We should somehow display the error.

Also, it looks like there is always an attempt to login with an existing WordPress.com account, i.e., with an account that is already linked to a social account. But in the signup step, shouldn't we be trying to _create_ a new user account? I'm confused here.

Reply:
Oh well, we should totally handle this then.

What I was thinking is to change the `finally` to a `then` and add a `catch` that handles the error and show something to the user.

Reply:
I'm able to reproduce this with "Continue with Apple". I sign in to Apple with "hide my email", it generates a _privaterelay.apple.com_ email with random-ish user name, and an account with username like `2czbhgtmcp` is created in the `/users/social/new` REST call. But a moment later, sending the same parameters to `/wp-login?action=social-login-endpoint` fails with `unknown_user`.

It fails silently, the signup continues without obvious errors. But the intent of this PR fails: the app state is still logged out.

I think that instead of doing social login, you want to do something else here. The social login has already succeded, and the `/users/social/new` response contains an OAuth `bearer_token`. To fully log the user in, i.e., to get the full set of WordPress.com auth cookies, you don't need to do a second social login. You can use the `bearer_token` to perform the Wordpress.com full login.

Currently the signup framework does this with the `WpcomLoginForm` component that creates a hidden `form`, with the bearer token as one of the fields, auto-submits it to `wordpress.com/wp-login.php`, and the form also has `redirect_to` parameter that specifes where to continue after login.

This is how the start flow should redirect to another place.

Reply:
I see that we get the `bearer_token` [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/signup/step-actions/index.js#L890)

So, I see that the signup framework already uses `WpcomLoginForm` as you said. The problem in our case is that social logins do not correctly set the `bearer_token` for the form to be created and submitted?

Reply:
The `/users/social/new` endpoint always returns a valid OAuth token in the `bearer_token` field, just like non-social `/users/new`. And the `responseHandler` function you're linking to always stores the token into signup dependencies.

So if the existing system doesn't work and you needed to patch it extra in this PR, then maybe the `WpcomLoginForm` is not created and submitted at the right time?

The `WpcomLoginForm` is rendered in the top-level `Signup` component, and is rendered as soon as a valid token appears in `this.state.bearerToken`. That state is set by the `handleLogin` function. Which is called only at signup completion, before redirecting you to the post-signup destination.

The P2 flow manually calls `handleLogin` right after the `user` step, before continuing to following steps. That means there is a full page reload after the `user` step and the rest of the flow is fully logged in.